### ISSUE TITLE
fisherman doesn't add local plugins to the fishfile

### ISSUE THREAD

**Issue by laughedelic** [fisherman/fisherman#242]
I have two local plugins that I'm installing with fisherman. They both get installed fine (symlinks created, functions work), but neither of them is listed in the fishfile. Here are my observations:

- if the plugin has name starting with a dot (e.g. `.fish`), it installs, but is not listed in the fishfile or `fisher ls`
- if it has a name without a dot, it installs, is listed in `fisher ls`, but not in the fishfile
- if it has a name without a dot and is a git repo with a remote set, it installs and is listed in both places, but differently:

```fish
> fisher i ~/dev/laughedelic/fish
INFO Installing dev/laughedelic/fish
OKAY Done in 144ms

> fisher ls
@ fish			  brew			  getopts		  host_info		  menu			  spin
* simple		  errno			  git_util		  humanize_duration	  pwd_info		  z
  anicode		  get			  grc			  last_job_id		  pwd_is_home

> cat ~/.config/fish/fishfile
fisherman/anicode
oh-my-fish/plugin-brew
oh-my-fish/plugin-errno
laughedelic/.fish.git # <-- same as the remote (https://github.com/laughedelic/.fish.git)
fisherman/get
fisherman/getopts
fisherman/git_util
oh-my-fish/plugin-grc
fisherman/host_info
fisherman/humanize_duration
fisherman/last_job_id
fisherman/menu
fisherman/pwd_info
fisherman/pwd_is_home
fisherman/simple
fisherman/spin
fisherman/z
```

**Comment by bucaran** [comment]
@laughedelic Thanks, I think this coincides with some of my observations too.

I think the correct behavior should be:

(1) to behave as (2), that is, to *not* be added to the fishfile and for (3) I am not sure.

**Comment by laughedelic** [comment]
> to not be added to the fishfile

Why? This makes local plugins some kind of special plugins (_all plugins are equal, but some plugins are more equal than others_ 😄).
What if I use fisher sometimes interactively and sometimes just editing fishfile? Some of the local plugins will be in the list, but others no.

**Comment by bucaran** [comment]
I can't recall now, but there was a reason I went with no adding local plugins to the fishfile. I could experiment with this later, as I don't find this a critical problem, or even a problem at all. 

I could be wrong, so I promise I'll have a look at this later.

> What if I use fisher sometimes interactively and sometimes just editing fishfile? Some of the local plugins will be in the list, but others no.

You lost me here. I have no idea what you are saying. It could be my brain, but try again and describe your use case again. ELI5.

**Comment by bucaran** [comment]
> @laughedelic 
> if the plugin has name starting with a dot (e.g. .fish), it installs, but is not listed in the fishfile or fisher ls

I see. I think we could agree this is a feature. Otherwise there would be no way to hide plugins.

WRT to (1) and (2) I found the issue is in `__fisher_plugin_get_url_info`, which explains why:

> laughedelic/.fish.git # <-- same as the remote (https://github.com/laughedelic/.fish.git)

**Comment by laughedelic** [comment]
> You lost me here. I have no idea what you are saying.

I guess a little example can help:

```fish
> fisher ~/plugin1
# installs plugin1...

> fisher ls
@ plugin1

> cat ~/.config/fish/fishfile
# empty

> echo '~/plugin2' >> ~/.config/fish/fishfile

> fisher
# installs plugin2...

> fisher ls
@ plugin1
@ plugin2

> cat ~/.config/fish/fishfile
~/plugin2

```

So the result is inconsistent handling of local plugins: one is in the fish file, another not. It's not obvious for user, what to expect further, on say, `fisher update`, also not obvious whether the first plugin got installed correctly or not. I believe, that if editing `fishfile` is promoted as a feature, it has to be consistent with `fisher ls`.

> I think we could agree this is a feature. Otherwise there would be no way to hide plugins.

What do you mean by hiding plugins and why it could be useful?

> WRT to (1) and (2) I found the issue is in __fisher_plugin_get_url_info

Great 👍

**Comment by bucaran** [comment]
@laughedelic Thanks, the example you've posted does seem like inconsistent behavior, so I'll fix this and release it with 2.4.

> What do you mean by hiding plugins and why it could be useful?

I don't know how it could be useful, but better to have a way to hide them than not. I recommend you do not start your plugin name with a `.`.

**Comment by bucaran** [comment]
@laughedelic Well, I couldn't repro this 😕 

```
$ echo "~/code/plugin2" >> ~/.config/fish/fishfile
$ fisher
INFO Installing 1 plugin/s
ARGH Fetching plugin2 → github.com/~/code/plugin2
ARGH There was an error cloning plugin2 or more plugin/s.
INFO Try using a namespace before the plugin name: xxx/plugin2
```

**Comment by bucaran** [comment]
I found this to be a regression introduced in #218.

**Comment by bucaran** [comment]
With https://github.com/fisherman/fisherman/commit/4583dd924be4ec85b53a14e530609f95c0c14841 the behavior is now consistent. If the plugins are bad / do not exist, they are never added to the fishfile.

The remaining issue is whether we should be adding plugins installed as local directories to the fishfile.

**Comment by bucaran** [comment]
I think we should not add local directory plugins to the fishfile, because if you move the fishfile to another computer it will not work. What about just making your plugin an actual repo and installing it with a universal URL or push it to github and installing it as a regular plugin?

**Comment by laughedelic** [comment]
> What about just making your plugin an actual repo and installing it with a universal URL or push it to github and installing it as a regular plugin?

It's no problem for me, but then it should be clearly stated in the docs, that local plugins are different from the ones installed from github and that editing fishfile is not the same as just using fisherman interactively. Because at the moment documentations states that these things are the same (and then it's confusing when they're actually not).

**Comment by bucaran** [comment]
@laughedelic Yep. But it's a very subtle difference. What would be the best way to state this?

**Comment by laughedelic** [comment]
Sorry, forgot to reply here. I'll think about it. But before that, I'm discovering new unexpected behaviour:

```
❯ fisher --version
fisherman version 2.6.2 ~/.config/fish/functions/fisher.fish

❯ fisher ls
@ ec2ls			  anicode		  choices		  get			  grc			  last_job_id		  pwd_is_home
@ fish			  await			  errno			  getopts		  host_info		  menu			  spin
* simple		  brew			  fzf			  git_util		  humanize_duration	  pwd_info		  z
```

At this point fishfile doesn't contain my local plugins. The to get autocomplete as mentioned in https://github.com/fisherman/fisherman/issues/258#issuecomment-217386488, I run

```fish
❯ fisher
INFO Installing 2 plugin/s
OKAY Fetch .fish.git github.com/laughedelic/.fish.git
ARGH Fetch ec2ls.git github.com/laughedelic/ec2ls.git
ARGH There was an error installing ec2ls.git or more plugin/s.
INFO Try using a namespace before the plugin name: xxx/ec2ls.git
```

- it tries to install my local plugins from their git remotes (which I think it shouldn't, if it treats them as local). it fails with ec2ls because I removed the repo from github.
- it adds this to my fishfile:

  ```
  laughedelic/ec2ls.git
  laughedelic/.fish.git
  ```

Another thing. Probably I should open a new issue for it.

```
❯ fisher rm choices
OKAY Done in 112ms

❯ fisher
INFO Installing 2 plugin/s
OKAY Done in 317ms
```

It installs `choices` and `get` and **adds them to the fishfile** "/

**Comment by bucaran** [comment]
I see a few issues in one here. But let's see:

> it tries to install my local plugins from their git remotes (which I think it shouldn't, if it treats them as local). 

Yeah, I agree, but I don't understand why I have never experienced this issue. I often install fisherman plugins, like `fin` or `menu` locally (even though they exist remotely) and work with them just fine. I can even update them locally as you'd expect. Can you crete a project or a way to repro it?

> ❯ fisher rm choices

I don't have the current state of your environment variables and I don't know what was in your fishfile. If I can't repro this, I can't fix it.

**Comment by laughedelic** [comment]
> Can you crete a project or a way to repro it?

I'll play with it more tonight and tell you if I have some stable way of reproducing it.

> I don't have the current state of your environment variables and I don't know what was in your fishfile. If I can't repro this, I can't fix it.

Sure, just tell me which vars to show you. About the fishfile: when I remove `choices`, it removes 

```
Markcial/Choices
fisherman/get
```

from the fishfile. They also disappear from `fisher ls`. When I run `fisher` without args after that, it installs them back and adds these lines back to the fishfile (and they appear again in `fisher ls`).

**Comment by bucaran** [comment]
> P.S. Everything same happens with the content of $fisher_dependency_count, I remove the plugin, those two disappear from the list, then I run fisher, they appear there again.

Wow. This is what I would like to repro. I tried a few scenarios, but everything is working normally.

**Comment by laughedelic** [comment]
Some more info: after `fisher rm choices`, it disappears from `$fisher_config` dir, but stays in `$fisher_cache` dir. I guess, it's expected behaviour. I tried to remove them from cache, but they still reappear! This is quite amazing... 

```
~/.cache/fisherman ❯ rm ~/.cache/fisherman/choices/
~/.cache/fisherman ❯ rm ~/.cache/fisherman/get/
~/.cache/fisherman ❯ fisher ls
@ fish			  await			  getopts		  host_info		  menu			  spin
* simple		  errno			  git_util		  humanize_duration	  pwd_info		  z
  anicode		  fzf			  grc			  last_job_id		  pwd_is_home

~/.cache/fisherman ❯ fisher
INFO Installing 2 plugin/s
OKAY Fetch choices github.com/fisherman/choices
OKAY Fetch get github.com/fisherman/get
OKAY Done in 2s 589ms
```

**Comment by laughedelic** [comment]
One detail that I just noticed. Before I had `Markcial/Choices`, after I cleaned the cache, it installed me `fisherman/choices`. But I still cannot get rid of it. 
@Markcial have you cursed this plugin? :trollface:

**Comment by Markcial** [comment]
I cursed it with pure awesomeness

Let me check what happens with local installs. Maybe there is a way to remove that plugin without explicitly removing it in the config folder

**Comment by laughedelic** [comment]
Tried both with `Markcial/choices` and `fisherman/choices`. Doesn't help. They get removed, but then reappear.

**Comment by bucaran** [comment]
@la

I think `choices` could be in the dependency list of some other plugins you have installed (a mistake) or they are in your fishfile.

fisherman uses your fishfile to check what to install and nothing more. Please cat and paste your fishfile here.

**Comment by bucaran** [comment]
@laughedelic 

I think `choices` could be in the dependency list of some other plugins you have installed (a mistake) or they are in your fishfile.

fisherman uses your fishfile to check what to install and nothing more. Please cat and paste your fishfile here.

**Comment by Markcial** [comment]
anicode uses choices as dependency, so here we have what is going on. 🔍   https://github.com/fisherman/anicode/blob/master/fishfile

**Comment by laughedelic** [comment]
OK:

```
❯ cat $fisher_bundle
fisherman/anicode
fisherman/await
oh-my-fish/plugin-errno
laughedelic/.fish.git
fisherman/fzf
fisherman/get
fisherman/getopts
fisherman/git_util
oh-my-fish/plugin-grc
fisherman/host_info
fisherman/humanize_duration
fisherman/last_job_id
fisherman/menu
fisherman/pwd_info
fisherman/pwd_is_home
fisherman/simple
fisherman/spin
fisherman/z

❯ fisher
INFO Installing 1 plugin/s
OKAY Done in 351ms

❯ cat $fisher_bundle
fisherman/anicode
fisherman/await
fisherman/choices
oh-my-fish/plugin-errno
laughedelic/.fish.git
fisherman/fzf
fisherman/get
fisherman/getopts
fisherman/git_util
oh-my-fish/plugin-grc
fisherman/host_info
fisherman/humanize_duration
fisherman/last_job_id
fisherman/menu
fisherman/pwd_info
fisherman/pwd_is_home
fisherman/simple
fisherman/spin
fisherman/z
```

But I think that the problem is here:

```
❯ cat $fisher_cache/.index | grep choices
choices	Configurable multiline choice selector with a prompt	github.com/fisherman/choices	7

❯ fisher rm choices
OKAY Done in 35ms

❯ cat $fisher_cache/.index | grep choices
choices	Configurable multiline choice selector with a prompt	github.com/fisherman/choices	7
```

I don't know how it got to 7, but anyway it got stuck there..

**Comment by bucaran** [comment]
7 is the number of stars.

**Comment by laughedelic** [comment]
Ah! I thought it's related to the count of dependent plugins! 😅

**Comment by bucaran** [comment]
Nope. The ref count is $fisher_dependency_count, echoing its contents here too please.

**Comment by Markcial** [comment]
wait a second, something is wrong, choices has 3 stars, so seems to be that choices is getting the stars from its parent dependency, anicode, the one that really has 7 stars.

**Comment by bucaran** [comment]
@Markcial Good catch! I doubt this is the problem, but that's certainly a bug you just found!

**Comment by laughedelic** [comment]
Strange:

```
❯ fisher rm choices
OKAY Done in 68ms

❯ echo $fisher_dependency_count
git_util pwd_info pwd_is_home host_info last_job_id humanize_duration simple last_job_id await fzf .fish.git fish spin get getopts getopts getopts getopts getopts getopts

❯ fisher
INFO Installing 1 plugin/s
OKAY Done in 324ms

❯ echo $fisher_dependency_count
git_util pwd_info pwd_is_home host_info last_job_id humanize_duration simple last_job_id await fzf .fish.git fish spin get getopts getopts getopts getopts getopts getopts getopts choices
```

I think, getopts wasn't multiplied before I cleaned cache..

**Comment by bucaran** [comment]
getopts will show up as many times as you have installed it, 1 for direct installs and n times for plugins that depend in it. The problem is not there either, since ref count is not involved when running `fisher` without args.

**Comment by laughedelic** [comment]
Just after that:

```
❯ fisher rm choices
OKAY Done in 65ms

❯ echo $fisher_dependency_count
git_util pwd_info pwd_is_home host_info last_job_id humanize_duration simple last_job_id await fzf .fish.git fish spin get getopts getopts getopts getopts getopts getopts

❯ fisher
INFO Installing 1 plugin/s
OKAY Done in 355ms

❯ echo $fisher_dependency_count
git_util pwd_info pwd_is_home host_info last_job_id humanize_duration simple last_job_id await fzf .fish.git fish spin get getopts getopts getopts getopts getopts getopts getopts choices
```

So getopts gets added every time choices reappears

**Comment by Markcial** [comment]
seems legit => https://github.com/fisherman/choices/blob/master/fishfile

**Comment by bucaran** [comment]
@

Do this:

```
fisher rm choices
__fisher_read_bundle_file < "$fisher_bundle"
```

**Comment by laughedelic** [comment]
Done:

```
❯ fisher rm choices
❯ __fisher_read_bundle_file < "$fisher_bundle"
fisherman/anicode
fisherman/await
oh-my-fish/plugin-errno
laughedelic/.fish.git
fisherman/fzf
fisherman/get
fisherman/getopts
fisherman/git_util
oh-my-fish/plugin-grc
fisherman/host_info
fisherman/humanize_duration
fisherman/last_job_id
fisherman/menu
fisherman/pwd_info
fisherman/pwd_is_home
fisherman/simple
fisherman/spin
fisherman/z
```

Didn't help. Also getopts is now listed as a local plugin for some reason:

```
❯ fisher ls
@ fish			  anicode		  errno			  git_util		  humanize_duration	  pwd_info		  z
@ getopts		  await			  fzf			  grc			  last_job_id		  pwd_is_home
* simple		  choices		  get			  host_info		  menu			  spin
```

**Comment by bucaran** [comment]
@la How did you upgrade to fisherman from ~1.5 and what version are you currently running?

List this stuff too:

```
ls -la $fisher_config
ls -la $fisher_cache
```

**Comment by laughedelic** [comment]
> How did you upgrade to fisherman from ~1.5

I don't remember already. Followed the instructions from the chat. I forgot to add lines to the fish config, but then added them.

> what version are you currently running

```
❯ fish --version
fish, version 2.2.0

❯ fisher --version
fisherman version 2.6.3 ~/.config/fish/functions/fisher.fish
```

That stuff:

```
❯ ls -la $fisher_config
total 16
drwxr-xr-x  22 laughedelic  staff  748 May  6 19:36 .
drwxr-xr-x   9 laughedelic  staff  306 May  1 18:35 ..
drwxr-xr-x   8 laughedelic  staff  272 May  6 17:00 .fish.git
drwxr-xr-x  12 laughedelic  staff  408 May  5 14:27 anicode
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 await
drwxr-xr-x  12 laughedelic  staff  408 May  6 19:36 choices
drwxr-xr-x   7 laughedelic  staff  238 Apr 28 22:48 errno
lrwxr-xr-x   1 laughedelic  staff   39 May  6 17:12 fish -> /Users/laughedelic/dev/laughedelic/fish
drwxr-xr-x   8 laughedelic  staff  272 May  6 16:50 fzf
drwxr-xr-x  13 laughedelic  staff  442 May  6 19:20 get
lrwxr-xr-x   1 laughedelic  staff   43 May  6 19:34 getopts -> /Users/laughedelic/.cache/fisherman/getopts
drwxr-xr-x  11 laughedelic  staff  374 May  6 16:50 git_util
drwxr-xr-x   7 laughedelic  staff  238 Apr 26 13:16 grc
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 host_info
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 humanize_duration
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 last_job_id
drwxr-xr-x   7 laughedelic  staff  238 Apr 28 22:48 menu
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 pwd_info
drwxr-xr-x  11 laughedelic  staff  374 May  6 01:09 pwd_is_home
drwxr-xr-x  10 laughedelic  staff  340 May  6 01:09 simple
drwxr-xr-x  13 laughedelic  staff  442 May  6 19:20 spin
drwxr-xr-x  14 laughedelic  staff  476 Apr 30 20:50 z

❯ ls -la $fisher_cache
total 32
drwxr-xr-x  24 laughedelic  staff   816 May  6 18:41 .
drwxr-xr-x   6 laughedelic  staff   204 May  5 18:54 ..
-rw-r--r--@  1 laughedelic  staff  6148 May  6 18:41 .DS_Store
drwxr-xr-x   8 laughedelic  staff   272 May  6 16:49 .fish.git
-rw-r--r--   1 laughedelic  staff  5162 May  6 17:03 .index
drwxr-xr-x  12 laughedelic  staff   408 May  6 01:09 anicode
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:42 await
drwxr-xr-x   6 laughedelic  staff   204 May  6 01:09 brew
drwxr-xr-x  12 laughedelic  staff   408 May  6 18:41 choices
drwxr-xr-x   7 laughedelic  staff   238 May  6 01:09 errno
drwxr-xr-x   8 laughedelic  staff   272 May  5 22:41 fzf
drwxr-xr-x  13 laughedelic  staff   442 May  6 18:41 get
drwxr-xr-x  11 laughedelic  staff   374 May  6 19:34 getopts
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 git_util
drwxr-xr-x   7 laughedelic  staff   238 May  6 01:09 grc
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 host_info
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 humanize_duration
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 last_job_id
drwxr-xr-x   7 laughedelic  staff   238 May  6 01:09 menu
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 pwd_info
drwxr-xr-x  11 laughedelic  staff   374 May  5 22:40 pwd_is_home
drwxr-xr-x  10 laughedelic  staff   340 May  5 22:40 simple
drwxr-xr-x  13 laughedelic  staff   442 May  6 01:09 spin
drwxr-xr-x  14 laughedelic  staff   476 May  6 01:09 z
```

**Comment by bucaran** [comment]
This is extremely weird ↓

```
lrwxr-xr-x   1 laughedelic  staff   43 May  6 19:34 getopts -> /Users/laughedelic/.cache/fisherman/getopts
```

**Comment by bucaran** [comment]
I am sorry, but I can't explain this. Can you remove all the plugins and then install choices and run `fisher` again?

**Comment by laughedelic** [comment]
Am I doing something wrong?

```
❯ cat $fisher_bundle | fisher rm
/Users/laughedelic/.z will be left intact...
to remove, simply delete it
OKAY Done in 715ms
fish: Unknown command 'grc.wrap'
in function 'cat'
	called on line 1426 of file ~/.config/fish/functions/fisher.fish
	with parameter list '/Users/laughedelic/.config/fisherman/fish/.git/config'

in function '__fisher_plugin_get_url_info'
	called on line 286 of file ~/.config/fish/functions/fisher.fish
	with parameter list '-- /Users/laughedelic/.config/fisherman/fish'

in function 'fisher'
	called on standard input
	with parameter list 'rm'

fish: Unknown command 'grc.wrap'
in function 'cat'
	called on line 1426 of file ~/.config/fish/functions/fisher.fish
	with parameter list '/Users/laughedelic/.config/fisherman/fish/.git/config'

in function '__fisher_plugin_get_url_info'
	called on line 2 of file ~/.config/fish/functions/fisher.fish
	with parameter list '-- /Users/laughedelic/.config/fisherman/fish'

in command substitution
	called on line 0 of file ~/.config/fish/functions/fisher.fish

in function 'fisher'
	called on standard input
	with parameter list 'rm'
```

**Comment by bucaran** [comment]
To remove all the plugins:

```
fisher l | fisher rm
```

**Comment by laughedelic** [comment]
After all this it doesn't reappear. One thing that I noticed on the way. When there are no plugins installed (empty fishfile), calling just `fisher` shows the usage message instead of the usual behaviour.

**Comment by bucaran** [comment]
@laughedelic Yep, that's correct.

**Comment by laughedelic** [comment]
LOL. It reappears again!

**Comment by bucaran** [comment]
It should be very easy to identify the bug though, since you have no plugins.

Please share as much information as possible.

**Comment by laughedelic** [comment]
It's a nightmare... 😩 
I do

```
> fisher ls | fisher rm
 will be left intact...                   # who? what?
to remove, simply delete it
OKAY Done in 628ms
> rm -rf $fisher_cache
```

At this moment everything is empty: fishfile, `$fisher_cache` and `$fisher_config` dirs and the `$fisher_dependency_count` var.

Then I add following plugins to the fishfile:

```
fisherman/menu
fisherman/anicode
fisherman/simple
fisherman/spin
fisherman/z
```

Because I want them. The I run `fisher` and get a bunch of errors:

```
> fisher
INFO Installing 5 plugin/s
OKAY Fetch simple github.com/fisherman/simple
OKAY Fetch menu github.com/fisherman/menu
OKAY Fetch z github.com/fisherman/z
OKAY Fetch anicode github.com/fisherman/anicode
OKAY Fetch spin github.com/fisherman/spin
INFO Installing 11 dependencies
OKAY Fetch getopts github.com/fisherman/getopts
cp: /Users/laughedelic/.config/fisherman/getopts: File exists
cp: /Users/laughedelic/.cache/fisherman/getopts: unable to copy extended attributes to /Users/laughedelic/.config/fisherman/getopts: No such file or directory
cp: /Users/laughedelic/.config/fisherman/getopts/.git: No such file or directory
cp: /Users/laughedelic/.cache/fisherman/getopts/.git: unable to copy extended attributes to /Users/laughedelic/.config/fisherman/getopts/.git: No such file or directory
cp: /Users/laughedelic/.config/fisherman/getopts/.git/branches: No such file or directory
cp: /Users/laughedelic/.cache/fisherman/getopts/.git/branches: unable to copy extended attributes to /Users/laughedelic/.config/fisherman/getopts/.git/branches: No such file or directory
cp: /Users/laughedelic/.config/fisherman/getopts/.git/config: No such file or directory
.... # a lot more like this
```

**Comment by laughedelic** [comment]
Oh crap!!! I'm sorry @bucaran! All this choices thing was coming from the anicode! it depends on choices... 😓

**Comment by bucaran** [comment]
>  will be left intact...                   # who? what?

This is `z`. I think @jethrokuan should be able to comment on that.

**Comment by bucaran** [comment]
You reacted with thumbs-down to your own post, that's so funny.
<img width="365" alt="screen shot 2016-05-07 at 3 29 38 am" src="https://cloud.githubusercontent.com/assets/8317250/15082656/fb93fa94-1403-11e6-9116-1d8eb4ea5ab8.png"> XD

**Comment by jethrokuan** [comment]
@laughedelic you need to have conf.d support for that to appear properly

**Comment by laughedelic** [comment]
> You reacted with thumbs-down to your own post, that's so funny.

I'm angry that I wasted so much of yours and my time because of such a stupid thing.

Nevertheless, I still see those `cp` error messages if I add those 5 plugins to the fishfile and run `fisher`.

**Comment by laughedelic** [comment]
> you need to have conf.d support for that to appear properly

@jethrokuan you mean [this](https://github.com/fisherman/fisherman#1-what-is-the-required-fish-version)? I have it 👌

**Comment by jethrokuan** [comment]
@laughedelic yeah that. If you had it then the error message wouldn't appear like that if `z` installed correctly.

**Comment by bucaran** [comment]
@laughedelic You did not waste my time at all. From this thread ~~I~~ @Markcial was able to discover a bug in the index that I am fixing right now and also found out about those nasty cp error messages which I'll get to after I am done with the index. 

In the mean time, try removing everything once again and see if you can reproduce the `cp` errors with less plugins and let me know.

**Comment by laughedelic** [comment]
@bucaran just did it. It fails with these two:

```
fisherman/anicode
fisherman/spin
```

@jethrokuan I don't understand it. If I have `z` installed and try to remove it, I see

```
❯ fisher rm z
/Users/laughedelic/.z will be left intact...
to remove, simply delete it
OKAY Done in 128ms
```

Is it supposed to be like that?

**Comment by jethrokuan** [comment]
@laughedelic yup!

initially you had
```
> fisher ls | fisher rm
 will be left intact...                   # who? what?
to remove, simply delete it
OKAY Done in 628ms
> rm -rf $fisher_cache
```

**Comment by bucaran** [comment]
Index problem fixed btw.

**Comment by bucaran** [comment]
@laughedelic What's your OS?

**Comment by bucaran** [comment]
I can't repro any of the errors :(

<img width="1018" alt="screen shot 2016-05-07 at 4 10 14 am" src="https://cloud.githubusercontent.com/assets/8317250/15083695/a013cdc4-1409-11e6-9c5f-d9bb97d06bd5.png">

**Comment by bucaran** [comment]
@jethrokuan Do you know of any weird differences between:

cp -rf and cp -Rf 

between Linux and OS X?

**Comment by laughedelic** [comment]
> What's your OS?

OS X

> I can't repro any of the cp errors 

Is it my unlucky day? "_

**Comment by bucaran** [comment]
@laughedelic If you are able to repro this, please create a new issue to track that.

**Comment by jethrokuan** [comment]
@bucaran `cp -r` and `cp -R` do the same thing on my OS

**Comment by bucaran** [comment]
@laughedelic Any updates on this? Maybe we can close this and create a new issue for the `cp` error, if you can repro it.

**Comment by laughedelic** [comment]
> Any updates on this?

@bucaran I've removed all plugins and did `fisher self-uninstall`, installed it again and added those two plugins to the fishfile — it worked fine and I haven't seen these cp errors "_

This issue got too far away from the original topic. The behaviour with installing local plugins is still inconsistent. Let's close this one and open a new issue for that?

Also, `fisher` without args with an empty fishfile shows usage message. I think it should behave the same as with a non-empty fishfile. https://github.com/fisherman/fisherman/issues/242#issuecomment-217516560

**Comment by bucaran** [comment]
> Also, fisher without args with an empty fishfile shows usage message. I think it should behave the same as with a non-empty fishfile. #242 (comment)

Yeah, I kind of agree with this. I want to do that, but also show a special message. I should push that in the next patch.
